File: go-backend/internal/collector/parsers/paloalto.go
package parsers

import (
	"regexp"
	"strconv"
	"strings"
	"time"
)

var paPrefixRe = regexp.MustCompile(`^\d+,\d{4}/`)

const pgIntMax = 2147483647

func init() {
	Register(&PaloAltoParser{})
}

// PaloAltoParser handles PAN-OS CSV syslog (TRAFFIC and THREAT types).
type PaloAltoParser struct{}

func (p *PaloAltoParser) ParserID() string  { return "paloalto" }
func (p *PaloAltoParser) DeviceType() string { return "firewall" }

func (p *PaloAltoParser) CanParse(line string) bool {
	check := line
	if len(check) > 20 {
		check = check[:20]
	}
	if !paPrefixRe.MatchString(check) {
		return false
	}
	parts := strings.SplitN(line, ",", 6)
	if len(parts) < 5 {
		return false
	}
	lt := parts[3]
	return lt == "TRAFFIC" || lt == "THREAT"
}

func (p *PaloAltoParser) Parse(line string) (map[string]interface{}, error) {
	fields := strings.Split(line, ",")
	if len(fields) < 36 {
		return nil, nil
	}

	eventTime := parsePATime(fields[1])
	if eventTime.IsZero() {
		eventTime = time.Now().UTC()
	}

	logType := fields[3]
	if logType == "TRAFFIC" {
		return parseTraffic(fields, eventTime), nil
	}
	if logType == "THREAT" {
		m := parseThreat(fields, eventTime)
		// URL Filtering logs are THREAT logs with subtype "url".
		// Augment (never replace) the THREAT output with web fields so the
		// existing LiveEvent production stays untouched.
		if strings.EqualFold(strings.TrimSpace(fields[4]), "url") {
			augmentURLFiltering(m, fields)
		}
		return m, nil
	}
	return nil, nil
}

// parseTraffic builds a normalised map from a PAN-OS TRAFFIC log line.
func parseTraffic(f []string, eventTime time.Time) map[string]interface{} {
	m := map[string]interface{}{
		"event_time":     eventTime,
		"source_ip":      f[7],
		"destination_ip": f[8],
	}

	setIntField(m, "source_port", f[24])
	setIntField(m, "destination_port", f[25])

	if f[29] != "" {
		m["protocol"] = strings.ToUpper(f[29])
	}

	m["action"] = normalizeAction(f[30])

	setIntField(m, "bytes_out", f[31])
	setIntField(m, "bytes_in", f[32])

	// Duration: PAN-OS reports seconds; convert to ms, cap at pgIntMax.
	if dur, err := strconv.Atoi(f[33]); err == nil {
		ms := dur * 1000
		if ms > pgIntMax {
			ms = pgIntMax
		}
		m["duration_ms"] = ms
	}

	setIntField(m, "packets_out", f[34])
	setIntField(m, "packets_in", f[35])

	if f[14] != "" {
		m["application"] = f[14]
	}
	if f[11] != "" {
		m["reason"] = f[11]
	}

	extractNAT(m, f)
	return m
}

// parseThreat builds a normalised map from a PAN-OS THREAT log line.
func parseThreat(f []string, eventTime time.Time) map[string]interface{} {
	m := map[string]interface{}{
		"event_time":     eventTime,
		"source_ip":      f[7],
		"destination_ip": f[8],
	}

	setIntField(m, "source_port", f[24])
	setIntField(m, "destination_port", f[25])

	if f[29] != "" {
		m["reason"] = f[29] // threat_id
	}

	m["action"] = normalizeThreatAction(f[30])

	if len(f) > 36 && f[36] != "" {
		m["application"] = f[36] // category
	}

	if len(f) > 37 && f[37] != "" {
		m["health_status"] = mapSeverity(f[37]) // severity
	}

	if len(f) > 38 && f[38] != "" {
		m["service"] = f[38] // direction
	}

	if f[29] != "" {
		m["protocol"] = strings.ToUpper(f[29])
	}

	extractNAT(m, f)
	return m
}

// augmentURLFiltering adds web-transaction fields from a PAN-OS URL Filtering
// log (THREAT subtype "url") on top of the standard THREAT map.
// Field positions follow the PAN-OS URL log layout:
// [31]=misc (the URL, often quoted), [35]=category, [41]=content type,
// [46]=user agent, [49]=referer, [56]=HTTP method. Trailing fields are only
// read when present so shorter lines still parse.
func augmentURLFiltering(m map[string]interface{}, f []string) {
	m["log_subtype"] = "url"

	if url := trimQuotes(f[31]); url != "" {
		m["url"] = url
		if host := hostFromURL(url); host != "" {
			m["host"] = host
		}
	}
	if len(f) > 35 {
		if cat := trimQuotes(f[35]); cat != "" {
			m["url_category"] = cat
		}
	}
	if len(f) > 41 {
		if ct := trimQuotes(f[41]); ct != "" {
			m["content_type"] = ct
		}
	}
	if len(f) > 46 {
		if ua := trimQuotes(f[46]); ua != "" {
			m["user_agent"] = ua
		}
	}
	if len(f) > 49 {
		if ref := trimQuotes(f[49]); ref != "" {
			m["referer"] = ref
		}
	}
	if len(f) > 56 {
		if method := trimQuotes(f[56]); method != "" && method != "unknown" {
			m["http_method"] = strings.ToUpper(method)
		}
	}
}

// trimQuotes strips surrounding double quotes and whitespace from a CSV field.
func trimQuotes(s string) string {
	return strings.Trim(strings.TrimSpace(s), `"`)
}

// hostFromURL extracts the host portion from a URL that may or may not
// include a scheme, path, or port.
func hostFromURL(url string) string {
	host := url
	if idx := strings.Index(host, "://"); idx >= 0 {
		host = host[idx+3:]
	}
	if idx := strings.IndexByte(host, '/'); idx >= 0 {
		host = host[:idx]
	}
	// Strip port (IPv4/hostname only; leave bracketed IPv6 untouched).
	if !strings.HasPrefix(host, "[") {
		if idx := strings.IndexByte(host, ':'); idx >= 0 {
			host = host[:idx]
		}
	}
	return host
}

// extractNAT populates NAT fields only when they differ from the original IPs.
func extractNAT(m map[string]interface{}, f []string) {
	if len(f) < 27 {
		return
	}

	srcIP, _ := m["source_ip"].(string)
	dstIP, _ := m["destination_ip"].(string)

	natSrc := f[9]
	natDst := f[10]

	if natSrc != "" && natSrc != "0.0.0.0" && natSrc != srcIP {
		m["nat_source_ip"] = natSrc
		if len(f) > 26 {
			setIntField(m, "nat_source_port", f[26])
		}
	}

	if natDst != "" && natDst != "0.0.0.0" && natDst != dstIP {
		m["nat_destination_ip"] = natDst
		if len(f) > 27 {
			setIntField(m, "nat_destination_port", f[27])
		}
	}
}

// normalizeAction maps PAN-OS TRAFFIC actions to canonical form.
func normalizeAction(raw string) string {
	low := strings.ToLower(strings.TrimSpace(raw))
	switch low {
	case "allow", "allowed":
		return "allow"
	case "deny", "denied":
		return "deny"
	case "drop", "dropped", "drop-all-packets":
		return "drop"
	}
	if strings.HasPrefix(low, "reset") {
		return "reset"
	}
	if low == "" {
		return "unknown"
	}
	return low
}

// normalizeThreatAction maps PAN-OS THREAT actions to canonical form.
func normalizeThreatAction(raw string) string {
	low := strings.ToLower(strings.TrimSpace(raw))
	switch low {
	case "alert":
		return "alert"
	case "allow", "allowed":
		return "allow"
	case "block", "blocked":
		return "deny"
	case "deny", "denied":
		return "deny"
	case "drop", "dropped", "drop-all-packets":
		return "drop"
	}
	if strings.HasPrefix(low, "reset") {
		return "deny"
	}
	if low == "" {
		return "unknown"
	}
	return low
}

var severityMap = map[string]string{
	"critical":      "critical",
	"high":          "high",
	"medium":        "medium",
	"low":           "low",
	"informational": "info",
}

func mapSeverity(raw string) string {
	low := strings.ToLower(strings.TrimSpace(raw))
	if v, ok := severityMap[low]; ok {
		return v
	}
	return low
}

// parsePATime tries the two common PAN-OS timestamp formats.
func parsePATime(raw string) time.Time {
	raw = strings.TrimSpace(raw)
	if t, err := time.Parse("2006/01/02 15:04:05", raw); err == nil {
		return t.UTC()
	}
	if t, err := time.Parse("2006/01/02 15:04:05.000000", raw); err == nil {
		return t.UTC()
	}
	return time.Time{}
}

// setIntField parses a string as int and stores it in the map, capped at pgIntMax.
func setIntField(m map[string]interface{}, key, val string) {
	val = strings.TrimSpace(val)
	if val == "" {
		return
	}
	n, err := strconv.Atoi(val)
	if err != nil {
		return
	}
	if n > pgIntMax {
		n = pgIntMax
	}
	if n < 0 {
		n = 0
	}
	m[key] = n
}

File: archive/python-backend-final/collector/parsers/paloalto.py
"""
Palo Alto Networks TRAFFIC + THREAT log parser.

Palo Alto firewalls emit CSV-formatted syslog with a well-known field
order.  The TRAFFIC log type is the primary source of session metadata
(5-tuple, NAT, bytes, duration, app-id, action).  THREAT logs carry
threat intelligence (threat ID, category, severity, direction, Wildfire
verdict) and are parsed into the same normalised schema by reusing
existing columns:

  threat_id          → reason
  threat_category    → application
  severity           → health_status
  direction          → service
  wildfire verdict   → appended to reason as " [wildfire: X]"

References:
  - PAN-OS 10.x/11.x Syslog Field Reference (TRAFFIC + THREAT types)
  - Fields are comma-separated; the log type token is at index 3.
"""
from __future__ import annotations

import re
from datetime import datetime
from typing import Any, Dict, Optional

from collector.parsers.base import BaseParser

# ── Shared field positions (TRAFFIC + THREAT) ────────────────────────────────
_F_TYPE               = 3
_F_TIMESTAMP          = 1     # receive_time: "YYYY/MM/DD HH:MM:SS"
_F_SRC_IP             = 7
_F_DST_IP             = 8
_F_NAT_SRC_IP         = 9
_F_NAT_DST_IP         = 10
_F_RULE_NAME          = 11
_F_SRC_PORT           = 24
_F_DST_PORT           = 25
_F_NAT_SRC_PORT       = 26
_F_NAT_DST_PORT       = 27
_F_APPLICATION        = 14

# ── TRAFFIC-specific positions ───────────────────────────────────────────────
_F_PROTOCOL           = 29
_F_ACTION             = 30
_F_BYTES_SENT         = 31
_F_BYTES_RECEIVED     = 32
_F_SESSION_DURATION   = 33    # elapsed seconds
_F_PACKETS_SENT       = 34
_F_PACKETS_RECEIVED   = 35

# ── THREAT-specific positions ────────────────────────────────────────────────
_F_THREAT_ID          = 29    # threat/content name (same index as proto for TRAFFIC)
_F_THREAT_ACTION      = 30    # alert|allow|block|reset-*|drop (same index as TRAFFIC action)
_F_THREAT_CATEGORY    = 35
_F_THREAT_SEVERITY    = 36
_F_THREAT_DIRECTION   = 39
_F_THREAT_WILDFIRE    = 40

# Severity mapping (PAN-OS label → normalised)
_SEVERITY_MAP: Dict[str, str] = {
    "informational": "info",
    "low":           "low",
    "medium":        "medium",
    "high":          "high",
    "critical":      "critical",
}

# Quick prefix check — Palo Alto lines start with "1," or a digit then
# a comma, followed by a date-like token.
_PREFIX_RE = re.compile(r"^\d+,\d{4}/")


class PaloAltoParser(BaseParser):
    PARSER_ID = "paloalto"
    DEVICE_TYPE = "firewall"

    def can_parse(self, line: str) -> bool:
        if not _PREFIX_RE.match(line[:20]):
            return False
        # Confirm TRAFFIC or THREAT at field 3
        parts = line.split(",", 5)
        if len(parts) < 5:
            return False
        return parts[_F_TYPE].strip() in ("TRAFFIC", "THREAT")

    def parse(self, line: str) -> Optional[Dict[str, Any]]:
        fields = line.split(",")
        if len(fields) < 36:
            return None

        log_type = fields[_F_TYPE].strip()
        if log_type not in ("TRAFFIC", "THREAT"):
            return None

        event_time = _parse_pa_time(fields[_F_TIMESTAMP].strip())
        if event_time is None:
            return None

        if log_type == "THREAT":
            return self._parse_threat(fields, event_time)
        return self._parse_traffic(fields, event_time)

    # ── TRAFFIC ──────────────────────────────────────────────────────────────

    def _parse_traffic(
        self, fields: list, event_time: datetime,
    ) -> Dict[str, Any]:
        action_raw = _safe(fields, _F_ACTION).lower()
        action = _normalize_action(action_raw)

        result: Dict[str, Any] = {
            "event_time":       event_time,
            "source_ip":        _safe(fields, _F_SRC_IP),
            "destination_ip":   _safe(fields, _F_DST_IP),
            "source_port":      _safe_int(fields, _F_SRC_PORT),
            "destination_port": _safe_int(fields, _F_DST_PORT),
            "protocol":         _safe(fields, _F_PROTOCOL).upper() or None,
            "action":           action,
            "reason":           _safe(fields, _F_RULE_NAME) or None,
            "bytes_in":         _safe_int(fields, _F_BYTES_RECEIVED),
            "bytes_out":        _safe_int(fields, _F_BYTES_SENT),
            "packets_in":       _safe_int(fields, _F_PACKETS_RECEIVED),
            "packets_out":      _safe_int(fields, _F_PACKETS_SENT),
            "application":      _safe(fields, _F_APPLICATION) or None,
        }

        # Duration (seconds → ms), capped at PostgreSQL integer max
        dur = _safe_int(fields, _F_SESSION_DURATION)
        if dur is not None:
            result["duration_ms"] = min(dur * 1000, 2_147_483_647)

        # NAT fields
        self._extract_nat(fields, result)
        return result

    # ── THREAT ───────────────────────────────────────────────────────────────

    def _parse_threat(
        self, fields: list, event_time: datetime,
    ) -> Dict[str, Any]:
        action_raw = _safe(fields, _F_THREAT_ACTION).lower()
        action = _normalize_threat_action(action_raw)

        # Threat ID → reason
        threat_id = _safe(fields, _F_THREAT_ID) or None
        reason = threat_id

        # Wildfire verdict → append to reason
        wildfire = _safe(fields, _F_THREAT_WILDFIRE).lower() if len(fields) > _F_THREAT_WILDFIRE else ""
        if wildfire and wildfire != "unknown":
            suffix = f" [wildfire: {wildfire}]"
            reason = f"{reason}{suffix}" if reason else suffix.strip()

        # Severity → health_status
        severity_raw = _safe(fields, _F_THREAT_SEVERITY).lower()
        severity = _SEVERITY_MAP.get(severity_raw, severity_raw or None)

        # Direction → service
        direction = _safe(fields, _F_THREAT_DIRECTION) or None

        # Category → application
        category = _safe(fields, _F_THREAT_CATEGORY) or None

        result: Dict[str, Any] = {
            "event_time":       event_time,
            "source_ip":        _safe(fields, _F_SRC_IP),
            "destination_ip":   _safe(fields, _F_DST_IP),
            "source_port":      _safe_int(fields, _F_SRC_PORT),
            "destination_port": _safe_int(fields, _F_DST_PORT),
            "action":           action,
            "reason":           reason,
            "application":      category,
            "health_status":    severity,
            "service":          direction,
        }

        # NAT fields
        self._extract_nat(fields, result)
        return result

    # ── Shared NAT extraction ────────────────────────────────────────────────

    @staticmethod
    def _extract_nat(fields: list, result: Dict[str, Any]) -> None:
        nat_src = _safe(fields, _F_NAT_SRC_IP)
        nat_dst = _safe(fields, _F_NAT_DST_IP)
        if nat_src and nat_src != result.get("source_ip"):
            result["nat_source_ip"] = nat_src
            result["nat_source_port"] = _safe_int(fields, _F_NAT_SRC_PORT)
        if nat_dst and nat_dst != result.get("destination_ip"):
            result["nat_destination_ip"] = nat_dst
            result["nat_destination_port"] = _safe_int(fields, _F_NAT_DST_PORT)


# ── Helpers ──────────────────────────────────────────────────────────────────

def _safe(fields, idx) -> str:
    try:
        return fields[idx].strip()
    except (IndexError, AttributeError):
        return ""


def _safe_int(fields, idx) -> Optional[int]:
    v = _safe(fields, idx)
    if not v:
        return None
    try:
        return int(v)
    except (ValueError, TypeError):
        return None


def _parse_pa_time(s: str) -> Optional[datetime]:
    """Parse Palo Alto timestamp: 'YYYY/MM/DD HH:MM:SS'."""
    for fmt in ("%Y/%m/%d %H:%M:%S", "%Y/%m/%d %H:%M:%S.%f"):
        try:
            return datetime.strptime(s, fmt)
        except (ValueError, TypeError):
            continue
    return None


def _normalize_action(raw: str) -> str:
    """Map Palo Alto TRAFFIC action tokens to the normalised vocabulary."""
    if raw in ("allow", "allowed"):
        return "allow"
    if raw in ("deny", "denied"):
        return "deny"
    if raw in ("drop", "dropped", "drop-all-packets"):
        return "drop"
    if raw.startswith("reset"):
        return "reset"
    return raw or "unknown"


def _normalize_threat_action(raw: str) -> str:
    """Map Palo Alto THREAT action tokens to the normalised vocabulary.

    THREAT logs use "alert" as the default action — we keep it as-is
    because it's semantically different from allow/deny.
    """
    if raw == "alert":
        return "alert"
    if raw in ("allow", "allowed"):
        return "allow"
    if raw in ("block", "blocked"):
        return "deny"
    if raw.startswith("reset"):
        return "deny"
    if raw in ("drop", "dropped", "drop-all-packets"):
        return "drop"
    if raw in ("deny", "denied"):
        return "deny"
    return raw or "unknown"
